src/workerd/api/sockets.c++

```diff
@@ -179,15 +178,18 @@ jsg::Ref<Socket> connectImplNoOutputLock(
     KJ_CASE_ONEOF(str, kj::String) {
       // We need just the hostname part of the address, i.e. we want to strip out the port.
       // We do this using the standard URL parser since it will handle IPv6 for us as well.
-      auto record = JSG_REQUIRE_NONNULL(url::URL::parse(jsg::usv(kj::str("https://", str))),
+      // It also converts all of the various odd forms of IPv4 addresses.
+      // We use a `fake://` protocol to avoid the automatic port normalization that
+      // would otherwise occur when using `https://` or `http://` here.
+      auto address = kj::str("fake://", str);
+      auto parsed = JSG_REQUIRE_NONNULL(jsg::Url::tryParse(address.asPtr()),
           TypeError, "Specified address could not be parsed.");
-      auto& host = JSG_REQUIRE_NONNULL(record.host, TypeError,
-          "Specified address is missing hostname.");
-      // Note that there is an edge case here where the address containing `:443` will nonetheless
-      // parse to a record whose `port` is set to nullptr.
-      auto port = record.port.orDefault(443);
-      isDefaultFetchPort = port == 443 || port == 80;
-      domain = host.toStr();
+      auto host = parsed.getHostname();
+      auto port = parsed.getPort();
+      JSG_REQUIRE(host != ""_kj, TypeError, "Specified address is missing hostname.");
+      JSG_REQUIRE(port != ""_kj, TypeError, "Specified address is missing a port.");
+      isDefaultFetchPort = port == "443"_kj || port == "80"_kj;
+      domain = kj::str(host, ":", port);
```

Comment:
The `domain` shouldn't contain the port

Reply:
Reworked this a bit more to ensure that no matter how the address is passed in the host and port and normalized consistently by passing them through the URL parser.

Reply:
btw hostname includes host and port.

Reply:
`host` is hostname and port, isn't it?

```
> const h = new URL('http://example.org:81')
undefined
> h
URL {
  href: 'http://example.org:81/',
  origin: 'http://example.org:81',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'example.org:81',
  hostname: 'example.org',
  port: '81',
  pathname: '/',
  search: '',
  searchParams: URLSearchParams {},
  hash: ''
}
```

Reply:
well, this was bound to happen some time. glad it was you.